src/counter.c

```diff
@@ -0,0 +1,98 @@
+// Commands for counting edges on GPIO input pins
+//
+// Copyright (C) 2021  Adrian Keet <arkeet@gmail.com>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+
+#include <stdbool.h>
+#include "basecmd.h" // oid_alloc
+#include "board/gpio.h" // struct gpio_adc
+#include "board/irq.h" // irq_disable
+#include "board/misc.h" // timer_read_time
+#include "command.h" // DECL_COMMAND
+#include "sched.h" // DECL_TASK
+
+struct counter {
+    struct timer timer;
+    struct gpio_in pin;
+    uint32_t poll_ticks;
+    uint32_t sample_ticks, last_sample_time;
+    uint32_t count, last_count_time;
+    bool value          : 1;
+    bool sample_pending : 1;
```

Comment:
FWIW, the use of C bitmap fields should be avoided.  Their behaviour is not well defined by the C specifications, and they can cause subtle issues depending on how the compiler implements them.  Indeed, they may have different implementations on AVR vs ARM vs PRU, etc.

So, as an example, were `sample_pending` to be updated outside of an `irq_enable(); ... ; irq_disable();` block, it could subtly add a read-modify-write operation to `value` and possibly introduce a race-condition.  I don't see anything in your code that could cause a problem, but I fear it's a "land mine" for future developers.  IMO, best to avoid C bitmaps.  Being explicit with flag manipulation is only a couple of extra lines of code and doesn't have these subtle issues.